Design Tools: Add a Position block support (including sticky), decoupled from layout

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Related

* https://github.com/WordPress/gutenberg/issues/30121
* https://github.com/WordPress/gutenberg/pull/44723
* https://github.com/WordPress/gutenberg/discussions/46032
* Will require the following to fix placement of toolbar controls while a block is selected: https://github.com/WordPress/gutenberg/pull/46565

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes: #30121 (note this PR is still a WIP, so the "fixes" label here is simply to associate this PR with the open issue)

This PR is an alternative to https://github.com/WordPress/gutenberg/pull/44723, and adds support for sticky position, decoupled from the Layout panel and whether or not a block has layout support. It explores ideas from the discussion in https://github.com/WordPress/gutenberg/discussions/46032

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Sticky / fixed position will be valuable for sticky headers and footers, and creative pattern applications. However, it is likely to be used infrequently, and needs to be de-emphasised in the UI. This PR is an exploration of the ideas in https://github.com/WordPress/gutenberg/discussions/46032, to see what the most minimal version of the feature might look like.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Take the ideas from #44723 and implement as a separate position block support
* Introduce the Position controls (design still to be finalised) under a separate Position panel in the inspector controls
* Opt the Group block in to using sticky position

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Note: sticky position will stick to the top of the direct parent block. If at first it appears not to be working, ensure that the parent block is tall enough / has enough content to be scrolled.

1. Open up the block editor, add a group block and put a paragraph inside it, and give the group block a background color
2. In the inspector controls, under Position, set the position to Sticky
3. Scroll down, and the block should stick to the top of the screen
4. Publish the post or page, and on the site frontend, the block should stick to the top as you scroll, and factor in the height of the logged-in admin bar when it is visible

## Screenshots or screencast <!-- if applicable -->

The position control situated under Position:

<img width="1115" alt="image" src="https://user-images.githubusercontent.com/14988353/205223008-57acb82c-3788-400c-b505-0af13cdb9883.png">

A screengrab of selecting the control from the Position panel:

![2022-12-02 16 34 11](https://user-images.githubusercontent.com/14988353/205222613-e422c3c1-7825-4eee-9d20-ece5217ca02a.gif)

andrewserong committed:
Try: Add Position support, decoupled from Layout

andrewserong committed:
Try a CustomSelectControl for the position input

andrewserong commented:
Update: based on discussion over in https://github.com/WordPress/gutenberg/discussions/46032, I thought I'd have a go at using a `CustomSelectControl` for the UI. Here's how it's currently looking:

https://user-images.githubusercontent.com/14988353/204980224-7e316684-2231-47f5-af11-fb57000d078a.mp4

andrewserong committed:
Try moving the control to a separate Position PanelBody

andrewserong commented:
Update: based on discussions over in https://github.com/WordPress/gutenberg/discussions/46032#discussioncomment-4282662, I thought I'd have a go at moving the control over to a separate `Position` `PanelBody`. Here's how it's looking now:

![2022-12-02 16 34 11](https://user-images.githubusercontent.com/14988353/205222613-e422c3c1-7825-4eee-9d20-ece5217ca02a.gif)

paaljoachim commented:
Thanks for working on this feature Andrew!

I am not updated on the discussion but my initial thought is that it would be nice to have this in a toggle. 
Position
Description based on toggle selection.
Toggle: Static - Sticky

jasmussen commented:
Took this for a spin, it's not bad:

![status](https://user-images.githubusercontent.com/1204802/205248610-653dde78-82ea-47e9-a556-ea25725eaac1.gif)

Of course if the split tabs experiment isn't enabled, the list can grow a bit long. Is it possible to show this position panel above the "advanced" panel?, i.e. at the end of the stack?

<img width="1392" alt="Screenshot 2022-12-02 at 09 22 49" src="https://user-images.githubusercontent.com/1204802/205248683-6b491260-6c5d-4a20-9d4e-2d36ffa8d311.png">

@paaljoachim We did consider a toggle. There's a longer discussion in #46032. One of the hopes here is we can start with an interface that can grow into supporting "Fixed" also.

@jameskoster @mtias could this serve as a starting point, that grows in the future as the position panel potentially becomes a toolspanel, and fixed is added?

andrewserong committed:
Add Position group, ensure Position panel always appears just before Advanced

andrewserong commented:
> Is it possible to show this position panel above the "advanced" panel?, i.e. at the end of the stack?

Good catch, and thanks for taking this for a spin!

I've pushed an update (https://github.com/WordPress/gutenberg/pull/46142/commits/338d16badcdf0849b9e7f41953dbe73f3a3fa89e) that explicitly places the `Position` panel directly before the `Advanced` panel in both contexts now 👍

| Normal state (appears after Border and before Advanced) | With the block inspector tabs experiment active (appears after Layout and before Advanced) |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/14988353/205559246-400262a5-524d-4991-bc0a-c962109df896.png" width="300" /> | <img src="https://user-images.githubusercontent.com/14988353/205559343-5b1b5f36-6033-4748-8b47-af743cacdcbb.png" width="300" /> |

jasmussen commented:
Thanks for your patience with this one. Current status:

![status](https://user-images.githubusercontent.com/1204802/205619622-ee484be3-fcb2-477b-af19-e02b40d74b5e.gif)

Some thoughts on launchability.

* The prominence in being attached to just group blocks is pretty low. That makes it pretty safe
* Sticky is easy, because it works well without any position values.
* The dropdown can expand, and we have some initial ideas on how to accomplish positioning for "Fixed" in the future.

That makes this a pretty low-footprint feature addition, that could probably only be made simpler by moving it to the "Advanced" panel, postponing the conversation on a "Position" panel.

I think the question then boils down to: is it useful to start with just sticky, or do we need fixed from the get-go after all?

* If we do need fixed, we need to think through more of the UI. 
* We likely need to create some codepens and explore how the various positional presets can work in a way that's useful.
* It might even be useful to consider some pattern designs that incorporate fixed positioning, and see how these might work in a branch. https://cloudup.com/ccO8mtV4UVn is one collection of designs where some can probably leverage "fixed".

I personally think we _can_ iterate (land things step by step) so long as we have some confidence in the end result. And so to answer the question: can we land "sticky" as the first step, it might be useful to ask some theme designers: how could you see yourself using it? Would it be useful in a pattern? CC: @WordPress/gutenberg-design  @WordPress/theme-team

andrewserong commented:
Thanks for the recap and thoughts about next steps @jasmussen!

Just a couple more things to add:

* For `fixed`, I'll need to do a bit more digging to work out how it'll play with the other elements in the editor (e.g. the layout header of the editor UI), as there'll be some styling conflicts there that the position logic will need to account for.
* When we set `fixed`, unlike `sticky`, the block will lose its width settings (e.g. the block doesn't default to full width). We don't yet have a container-level width control, so we probably need to figure out how adjusting width (or height) will work for fixed position blocks. Fortunately we have the `min width` support for Group blocks now, so height is probably okay, it's mostly width we'll need to figure out.
* Whichever way we go, if we decide to begin with this PR and just `sticky` support to begin with, there are still some issues with the block selection toolbar to resolve, which I'm investigating separately over in https://github.com/WordPress/gutenberg/pull/46085 and hope to get resolved soon 🤞

andrewserong committed:
Roll in fix for toolbar position

andrewserong commented:
I'm wrapping up for the year now. To keep this a working snapshot of the current progress toward a sticky position support, I've given this a rebase and rolled in one of the proposed fixes for the positioning of the block selection toolbar (https://github.com/WordPress/gutenberg/pull/46565). While there's still things that need to be tweaked in this PR I'm sure, this should be an accurate representation now of where we're currently at.

Here is a quick demo of the current state, and how to set a Header to be sticky by wrapping it in a Group block that is set at the root of the site, and then switching it over to "sticky":

https://user-images.githubusercontent.com/14988353/207780197-f0dad53e-35bc-4370-b1e4-fe7c58f6e268.mp4

In the above video:

* The theme has a Header template part at the root of the site.
* Select the Header template part, and wrap it in a Group block.
* Deselect `Inner blocks use content width` to make the block full width.
* With the root Group block selected, set the Position to Sticky.
* It appears to work fairly well so far

Some of the limitations of the current approach:

* Because `sticky` is set at the Group block level, this happens outside of the Header template part, so for a site with many different templates, each _template_ would need to have this set, rather than setting it within the Header template part. A possible longer-term solution might be to allow block supports at the template part level, however since this hasn't been done up until now, it might be non-trivial to attempt, and earlier discussion flagged that as a good thing to explore in follow-ups rather than for us to try to do that just yet.
* The output of the sticky position rules on the site front end factors in the logged in WP header, which works nicely at desktop viewport sizes. However, on mobile, it appears that WP header isn't sticky, so I'm not too sure what the ideal relationship between that header and a block that is set to the top position would be on mobile on a site frontend. This might need a little more digging.
* While the block toolbar position is improved thanks to the changes pulled in from #46565, it's still a little jittery, so will need some more digging, too, to come up with an ideal fix there.

I'll pick this up again when I'm back in the new year, but very happy if anyone wants to hack on this PR, branch from it, and/or help out with the related PR to look into block selection toolbar fixes (#46565). No worries if not, I'll happily resume from where I left off once I'm back.

Thanks again for all the feedback and collaboration on this one!

jameskoster commented:
> Because sticky is set at the Group block level, this happens outside of the Header template part, so for a site with many different templates, each template would need to have this set, rather than setting it within the Header template part

Imo this is fine. It should always be the template that defines stickiness, not the part itself. Otherwise we'd need to design a UI to override the inherent sticky behavior per template, and that starts to get very complicated. 

The flexibility can be useful too – on some templates you might want to include additional blocks in the sticky region (sale banner on a product category archive for example).

It's a shame that you have to wrap the template part in a Group as that isn't particularly discoverable, but we can probably solve that separately. It may only require something simple like a "Sticky Group" variation.

jasmussen commented:
> It's a shame that you have to wrap the template part in a Group as that isn't particularly discoverable, but we can probably solve that separately. It may only require something simple like a "Sticky Group" variation.

Can't it be a group inside the template part?

jameskoster commented:
I don't think so, because `position: sticky;` is always relative to its container. Since the Template Part block outputs a container in the markup the Group wouldn't stick. Crude example:


https://user-images.githubusercontent.com/846565/208473654-a1c1a8ad-7b5d-42be-b484-d09d5a073741.mp4

jasmussen commented:
Oh, I forgot that. Does the template part always output a container? I may have mentally grouped it with RBs. If yes, then if need be we could bring sticky to the template part itself, if it turns out succesful.

jameskoster commented:
It gets tricky if we do that. We learned before that folks expect template part edits to be global, but block properties are local to the parent document. This is why we had to remove background color options on template parts back in the day, it just gets way too confusing when you mix global / local settings.

jasmussen commented:
Indeed. I don't think it's the end of the world to wrap the template part in a group.

jameskoster commented:
Agreed, especially if we decide that a "Sticky" Group variation makes sense.

jasmussen commented:
If we are to have a single interface for sticky and fixed, potentially absolute in the future, and possibly an interface that might even expand beyond just the group, wouldn't that make a "sticky group" variation redundant?

jameskoster commented:
I don't think so. I see the variation serving two purposes:

1. In list view it's easier to identify a sticky region
2. Adding a sticky region from the inserter is easier and the functionality more discoverable

That said, it doesn't deserve the same prominence as other variations like Row and Stack, especially as Rows and Stacks can be sticky too :) So perhaps it's less of a variation in the traditional sense, and more of a treatment conditionally applied to containers based on `position`. 

Definitely not something to worry about here/now though :)

andrewserong commented:
Thanks for the discussion here, folks! I'm back from end of year AFK, so I'll start working on this one again. I've given it a rebase so that it's current with trunk now.

> This is why we had to remove background color options on template parts back in the day, it just gets way too confusing when you mix global / local settings.

Good to know the extra context here about the challenges of adding attributes to template parts, it sounds like we're aligned now that (at least in the short term), for folks creating sticky headers, they'll be wrapping their header template part in a Group block set to sticky, within their template. While not very discoverable, I do like that it makes for an explicit separation, and it means that whether or not the header is sticky is separate from how header patterns work. I.e. if someone has a sticky group block at the top of their template, they can choose to put any header pattern within it, rather than reaching for a "sticky header" pattern.

tellthemachines commented:
> While not very discoverable, I do like that it makes for an explicit separation, and it means that whether or not the header is sticky is separate from how header patterns work.

I agree that it's nice to have the separation of concerns so that the same header can be used as either sticky and non-sticky, but the big downside to that is having to wrap the header in a sticky container on each template we want it to be sticky in. It would be great to have a way to make it sticky across all template parts at once.

Would we be able to do something like set `display: contents` on the template part? That way it would be possible to have the sticky position wrapper _inside_ the template part and have it stick to the template part's container.

andrewserong commented:
> Would we be able to do something like set display: contents on the template part? That way it would be possible to have the sticky position wrapper inside the template part and have it stick to the template part's container.

Oh, thanks, TIL about `display: contents` — I've just had a quick hack around with adding `display: contents` as a styling rule to `.wp-block-template-part`, but it looks like a side effect of that was that it causes blocks to no longer be able to be selected in the editor canvas via clicking, and the area for the block within the site editor winds up being calculated as being the full height of the editor canvas (note the purple outline in the screenshot indicates that the editor thinks the header template part now extends to the bottom of the page). I'm not sure how much of that can be bypassed within the editor, and how much is due to browser implementation issues with `display: contents`? The [caniuse page](https://caniuse.com/css-display-contents) for it lists a few known issues 🤔

![image](https://user-images.githubusercontent.com/14988353/210484793-26937dd3-3d1a-4613-93d4-f57a444084eb.png)

I think it'd be worth it to do further exploration into the template part block separately to this PR, though. As you mention, it'd be great for users to be able to avoid having to wrap the template part on each of their templates if possible.

tellthemachines commented:
> I think it'd be worth it to do further exploration into the template part block separately to this PR, though. As you mention, it'd be great for users to be able to avoid having to wrap the template part on each of their templates if possible.

Yeah, definitely not a blocker for this PR!

andrewserong committed:
Remove no longer needed change to block popover

andrewserong committed:
Add scrollContainer to deps array

apeatling commented:
Tested this with Twentytwentythree, wrapping the header template part in a group block for the index template. It worked well, very cool! 

I do think the discoverability of needing to wrap a header template part in a group block is very low, but understand the reasoning. I can imagine that making a header of a site sticky is a primary use case for this.

![2023-01-04 09 40 42](https://user-images.githubusercontent.com/1464705/210616782-c7d0938f-d500-4d03-9ca7-bfdb6ef87baf.gif)

andrewserong committed:
Remove redundant PHP code

andrewserong commented:
Thanks for testing! I've moved this PR out of draft now that is feels like we're honing in on an MVP scope for the feature now.

jameskoster commented:
Nice this seems to be working well for me.

One detail I am on the fence about is whether we need the descriptions? The one for Sticky needs to clearly communicate the option more effectively than the "Sticky" label alone. I'm not sure it does that. Maybe it's just a case of refining it, but it's a tricky behaviour to describe :) An alternative option:

> The block will stick to the top of the window instead of scrolling

For the Static option, it might be more intuitive to simply present that as "Default" with no description?

apeatling commented:
> For the Static option, it might be more intuitive to simply present that as "Default" with no description?

I agree that default is better here. "Static" for those that don't have experience with CSS could easily mean "doesn't move", which compared with "Sticky" might be confusing.

A description line would be worth trying for each option outside of default.

andrewserong committed:
Switch Static to Default, display help text after selection, update text

tellthemachines commented on lib/block-supports/position.php:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	$style_attribute = _wp_array_get( $block, array( 'attrs', 'style' ), null );
+	$class_name      = wp_unique_id( 'wp-container-' );
```
Do we anticipate having to target children of positioned blocks with CSS, or are there others reason to use the layout support logic here as opposed to the mix of preset classnames and inline styles used for other block supports?

andrewserong replied:
Oh, good point! The approach I went with here predates the `WP_HTML_Tag_Processor`, without which it would have been tricky to append the style rules, but you're right, each of the generated rules should be able to happily exist as an inline value.

It's definitely worth a try, and would be a neater approach if it works cleanly 👍

tellthemachines commented on lib/block-supports/position.php:
```diff
@@ -0,0 +1,130 @@
+				'selector'     => $selector,
+				'declarations' => array(
+					'position' => $position_type,
+					'z-index'  => '10', // TODO: This hard-coded value should live somewhere else.
```
Could we associate this with a static classname so it could be attached to the [existing z-index scale](https://github.com/WordPress/gutenberg/blob/try/position-support-decoupled-from-layout/packages/base-styles/_z-index.scss#L1)?

andrewserong replied:
Oh, I like the idea of a classname approach. `z-index` could potentially work nicely with presets now that you mention it, as we could have a preset for the sticky level. I was planning to look further into how to manage z-indexes in a follow-up PR rather than this one, as it could be a bit fiddly to get right. Maybe as a first step for this PR, if we inject `z-index: 10` in the inline style along with the position / top value... then in a follow-up we can look at creating the preset?

One thing to note: because the z-index value is output to the site frontend, it needs to exist separately from the existing z-index scale, which is only present in the editor.

But I do like the idea of presets for z-index values, a bit like we have for font sizes.

tellthemachines replied:
Yeah, I think it's fine to leave the z-index refinements for a follow-up 👍

andrewserong commented:
Thanks for the feedback! Now that you mention it, yes, "Static" could sound quite a lot like "Fixed" to a user not familiar with the CSS values. I've done the following:

* Updated Static to Default
* Removed help text from Default
* Updated Sticky help text to read `The block will stick to the top of the window instead of scrolling`
* Set the help text to display while the value is selected — I think this might help a little to make it clearer what the current value is / means before someone clicks the drop-down.

Happy to revert any of that / tweak further. A nuance that's difficult to communicate is that sticky is sticky to the parent container rather than to the window, but I'm not sure how specific we want to get in the help text.

Also, does it look a bit empty if Default has no explanatory text at all? The MDN description for static is a bit dry (`The element is positioned according to the normal flow of the document.`), but I wonder if it's worth trying to describe it a bit — mostly I'm think of a user who hasn't seen the setting before and wonders what Default sitting under a Position panel means.

Here's how the UI is looking now:

| Dropdown open | Sticky selected |
| --- | --- |
| <img width="286" alt="image" src="https://user-images.githubusercontent.com/14988353/210929234-bd1c336e-95e4-4629-8f7d-a9842fc5d411.png"> | <img width="282" alt="image" src="https://user-images.githubusercontent.com/14988353/210929256-88451d58-8403-4a09-abe2-974646e80c4c.png"> |

tellthemachines commented on packages/block-editor/src/components/inspector-controls-tabs/settings-tab.js:
```diff
@@ -2,11 +2,13 @@
 const SettingsTab = ( { showAdvancedControls = false } ) => (
 	<>
 		<InspectorControls.Slot />
+		<PositionControls />
```
I'm not altogether sure how this works anymore 😅 but wouldn't this panel be added directly inside InspectorControls like in [Layout](https://github.com/WordPress/gutenberg/blob/try/position-support-decoupled-from-layout/packages/block-editor/src/hooks/layout.js#L197)?

andrewserong replied:
This is an interesting one — I tried it like that originally, but we needed to have deliberate control over where the Position panel is output in the list of panels, and using `<InspectorControls>` as the parent means that the order is a bit harder to predict. By setting it here explicitly, we ensure that it always appears at the bottom, directly before the advanced controls, rather than using the order in which components are imported to determine the rendering order.

tellthemachines reviewed (commented):
This is looking really close! It does make sense to present it as a separate control from layout. the questions below are mainly me trying to understand how everything fits together 😄 

One thing that felt a bit weird in testing: If it's a small block, the toolbar can completely overlay it once it's sticky:
<img width="651" alt="Screenshot 2023-01-06 at 3 36 36 pm" src="https://user-images.githubusercontent.com/8096000/210933844-d249e5f2-dff3-4f6a-ae23-7092acf05ff4.png">
(I had to add the border there to make sure the block was sticking correctly)

Apart from that it works very well ✅

andrewserong commented:
Thanks for taking this for another spin @tellthemachines! I'll try out switching to inline styles on Monday, I think that'll really simplify things, and it's good to avoid having to enqueue extra styles if we don't have to 👍

> One thing that felt a bit weird in testing: If it's a small block, the toolbar can completely overlay it once it's sticky:

Oh, interesting. I can only replicate that with the Group block's placeholder state. For me, if the block has real content in it (e.g. a paragraph), then the toolbar gets out of the way, even if the Group block is selected:

<img width="688" alt="image" src="https://user-images.githubusercontent.com/14988353/210937094-f6ee61c3-de61-4c27-bc00-7e45b95fe972.png">

I wonder if the measurement for the space that the block takes up isn't working as expected when the block is in its placeholder state 🤔... I'll dig in a little next week.

jameskoster commented:
> Set the help text to display while the value is selected

I'm not sure about this one. Generally help text would describe the control rather than the selected value. There might be a11y issues if we break that convention. If the option has a description that's probably adequate?

andrewserong commented:
> There might be a11y issues if we break that convention. If the option has a description that's probably adequate?

Okie-doke! I'll remove it 🙂

andrewserong committed:
Remove help text, update logic to gracefully handle illegal values

andrewserong committed:
Try ensuring sticky/fixed blocks always flip the toolbar position

andrewserong commented:
A quick update from today:

* I've removed the help text I added the other day, so the only help text is what's displayed in the drop-down list.
* I've had a go at resolving the issue that @tellthemachines encountered with toolbar positioning on very small blocks. This appears to be because the logic I'd used for flipping the position of the toolbar wasn't quite good enough for a sticky block that is selected _before_ it's scrolled off the edge of the screen. I've now updated the logic so that a sticky or fixed block will always flip the toolbar to be underneath the block if there isn't enough room above it, which appears to fix that up as far as I can tell. It adds a little more "knowledge" about blocks to the hook for the block toolbar, but feels a bit better to use, so I think it's probably worth it.
* In terms of the CSS output, I took another look at the code and remembered that I initially went with enqueueing / adding CSS rules via the style engine function instead of inlining them so that it's easier to add either combined rules or potentially media queries in the future, if it turns out we eventually need to support different positions based on viewport width. While we likely won't be implementing anything like that in the shorter-term, keeping the enqueued CSS (using a similar approach to the layout support) gives us a bit more flexibility to try things in follow-up PRs.

jameskoster reviewed (approved):
I think this one is in a good spot from the design perspective. It positions (excuse the pun) us nicely for future enhancements down the road, and is a solid base from which we can learn more about general usage.

tellthemachines commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -0,0 +1,376 @@
+);
+
+/**
+ * Override the default block element to add the layout styles.
```
Should it be "position" instead of "layout" in this and also in the comment above?

andrewserong replied:
Yes! I've updated these two comments.

tellthemachines commented on packages/block-editor/src/hooks/position.js:
```diff
@@ -0,0 +1,376 @@
+		const element = useContext( BlockList.__unstableElementContext );
+
+		// Higher specificity to override defaults in editor UI.
+		const positionSelector = `.wp-container-position-${ id }.wp-container-position-${ id }`;
```
Could we just use `wp-container-` here, to match the server side? (Ideally these styles and the layout ones would be consolidated under the same class but that's probably a task for later 😅 )

andrewserong replied:
Good call, I've updated these to match for now 👍

tellthemachines reviewed (commented):
> I've had a go at resolving the issue that @tellthemachines encountered with toolbar positioning on very small blocks.

Thanks for fixing this! Can confirm it's now working for me. I think I might have forgotten to mention that the block I was testing with was set to Row.

The code looks good, there's just one thing I noticed when testing with TT1. I'm not sure we should be showing Position controls at all for classic themes? Currently the panel appears but not the controls:

<img width="281" alt="Screenshot 2023-01-10 at 11 20 36 am" src="https://user-images.githubusercontent.com/8096000/211437123-1f0c88b8-9730-42e3-aeae-504fb47c91ab.png">

andrewserong commented:
Thanks for re-testing @tellthemachines! Good catch about the Position panel on classic themes, yes, the panel should be hidden altogether if position isn't opted-in... I'll fix it up along with the other comments you've noted 🙂

andrewserong committed:
Ensure position UI and style output is only applied on themes that support it, tidy up comments

andrewserong commented:
Update: in https://github.com/WordPress/gutenberg/pull/46142/commits/911771ff08901704e73257ae2cf77019fd54b0b0, I've updated the logic so that the whole Position panel is hidden when there is no position support in the theme. I've also updated the style output logic so that the fixed/sticky position styles are also not output if the theme doesn't have support — this way, if someone has a block set to sticky and switches to a theme that does not support it, the sticky styles aren't output. I think that's probably the best way to go about it, because if we don't restrict output, there'd be no way for a user to switch the sticky/fixed positioning off, if a block has it, and the theme doesn't.

andrewserong committed:
Clarify TODO comment for z-index values

tellthemachines reviewed (approved):
Thanks for the update! Re-testing confirms the Position panel no longer appears on classic themes, even if there's a positioned block in the content.

andrewserong commented:
Thanks for all the reviews and collaboration on this one, everyone!

> It positions (excuse the pun) us nicely for future enhancements down the road, and is a solid base from which we can learn more about general usage.

Agreed, I think we've honed in on a good initial set of features for now. I've opened up an issue in #47043 to list follow-up tasks, which should be a little easier to implement now that we've got the foundation in. Feel free to add / edit the list if anyone has other items to add!

youknowriad commented:
Hi folks, I missed the introduction of this block support but I have some high-level questions/thoughts about this.

At first, I'd agree that setting a layout and setting a position is two different things but there are strong relations between the two:

 - First thing I want to ask is why this is even an opt-in block support: I guess it's a safe start but setting position seems to be something that all blocks should have access to.
 - I've always considered the alignments we have now: left/right/wide/full as some properties of the "position" definition.
 - I think position depends on layout. In the sense that if the parent layout is "flex", it doesn't make sense to have "alignments" but does it make sense to have "sticky"?
 - Are we going to support "absolute" position at some point, in which case, the "position" object would have a different format (type, x, y).

Are these things you've discussed, what's the current approach and do we have a "plan" to go there?

jameskoster commented:
> I guess it's a safe start but setting position seems to be something that all blocks should have access to.

I agree, but the 'safe start' element is why we restricted this to Groups for now. It also enables the creation of sticky headers which was one of the driving forces in adding this functionality.

> Are we going to support "absolute" position at some point

I believe so, yes. There's lots of back-scroll on the design discussion [here](https://github.com/WordPress/gutenberg/discussions/46032). We explored how the control can support other position values (absolute and fixed), then rowed back to the simplest execution for sticky as a starting point.

andrewserong commented:
Thanks for adding the thoughts and questions @youknowriad!

> I agree, but the 'safe start' element is why we restricted this to Groups for now.

Yes, this gives us a much safer level of scope for testing and making further changes — once we've got a stabilised and fully fleshed out suite of features, we could look at switching over to more of an opt-out approach. For the moment, it's been very helpful being deliberate about the opt-in approach, especially with feedback on how `sticky` works where it's looking likely that we'll hide the sticky controls when the block is in a non-root position (more discussion here: https://github.com/WordPress/gutenberg/pull/47230#issuecomment-1396632624).

> I think position depends on layout. In the sense that if the parent layout is "flex", it doesn't make sense to have "alignments" but does it make sense to have "sticky"?

There's definitely a close relationship between layout and position, but after previous PRs that explored tying the implementation of position to layout, it became clear that it's helpful for there to be at least some separation between the two, which is how we settled on having it be its own support. It gives us more flexibility in being able to open up controls outside of layout (e.g. relative/absolute) in the future. However, it also doesn't box us into a corner in that we can always roll the controls into the Layout UI in the future if we need to.

But one of the other concerns is that Layout has already grown quite a lot in complexity, so to manage that, it's helpful to have a position support that uses a similar approach, but sits just off to the side.

> Are we going to support "absolute" position at some point, in which case, the "position" object would have a different format (type, x, y).

Yes, we considered that and it informed the shape of the position object. Absolute would be represented in the same way as the current sticky position, using `top`, `right`, `bottom` and `left` properties to match the CSS. So the following shapes are tacitly supported by the current PR:

```
// sticky (implemented in this PR)
{
  "type": "sticky",
  "top": "0px"
}

// absolute (supported by the structure in this PR, not yet implemented)
{
  "type": "absolute",
  "top": "-15px",
  "left": "-15px"
}
```

So, we don't _yet_ support relative/absolute output, etc, but the shape of the data will support it.

> Are these things you've discussed, what's the current approach and do we have a "plan" to go there?

The plan for the moment is to refine the controls for sticky positioning, with the objective to land it as part of 6.2. The work is being tracked in a follow-up issue of tasks here: https://github.com/WordPress/gutenberg/issues/47043

For future exploratory designs, there is discussion over in https://github.com/WordPress/gutenberg/discussions/46032, with some UI controls being explored further down, too.

In terms of next steps, I think once we're past the 6.2 feature freeze and have sticky positioning polished to the best place we can, it'd be good to start with a design-led approach to how the UI/UX should work for the other kinds of positioning, and then we could look at breaking down the tasks from there. In terms of the data model for attributes / structure of where the code lives, I think we landed on a pretty good approach in this PR that should set us up well for exploring those follow-ups.

andrewserong commented:
## Dev notes for 6.2

Note: if there's room, this dev note could potentially be combined with other design tools in a larger dev note. There is a bit of information to convey, though, so alternately it could be its own standalone note. Draft below:

---

### Sticky position block support

WordPress 6.2 adds a new position block support feature, beginning with support for sticky positioning, with the Group block opted-in by default. For themes using the `appearanceTools` feature in `theme.json`, the Group block will now be able to be set to "sticky".

The sticky positioning feature follows [the CSS behavior for `position: sticky`](https://developer.mozilla.org/en-US/docs/Web/CSS/position) which sets an element to stick to its immediate parent (in this case, the block's immediate parent) when the user scrolls down the page. Because it can be a potentially confusing feature to work with, for WordPress 6.2, the positioning controls will only be displayed for Group blocks at the root of the document. In order to create a sticky header in a site template, people building themes or sites can wrap a Header template part in a Group block and set that Group block to "sticky".

This is just the beginning of the position block support feature, and there are ideas for future enhancements, including rolling out sticky to non-root blocks, and improving the UX surrounding building sticky site headers. There is [an open issue](https://github.com/WordPress/gutenberg/issues/47043) in the Gutenberg repo to track progress beyond 6.2.

#### How to add sticky position support to a theme

There are two ways to add support for sticky position to a blocks theme. The simplest is to opt in to the `appearanceTools` setting, which automatically enables a number of design tools ([read more in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls)).

For themes that wish to have more granular control over which UI tools are enabled, the sticky position support can be opted into by setting `settings.position.sticky` to `true` in `theme.json`. For example:

```
{
	"settings": {
		"position": {
			"sticky": true
```

#### How does it work?

When a user sets a block to the sticky position via the drop-down list in the inspector controls, styling rules are output for `position: sticky`, a `top` position of `0px` (with an offset to account for the logged in admin bar), and a hard-coded `z-index` value of `10` to ensure that the sticky block sits above other content, while still working correctly in the admin UI. There is not yet support for controlling custom z-index values.

To support the output of these position properties, the `safe_style_css` filter [has been updated in core](https://core.trac.wordpress.org/ticket/57504) to support the `position`, `top`, `right`, `bottom`, `left`, and `z-index` properties.

The styling rules are output when the block is rendered, and a classname `is-position-sticky` is injected into the block markup. For themes wishing to add custom CSS that targets sticky positioned group blocks, this could be achieved by using a selector like `.wp-block-group.is-position-sticky`.


#### Why was support for non-root sticky blocks removed in Gutenberg?

While the feature was being developed in the Gutenberg plugin, sticky positioning was initially enabled in all hierarchies of blocks. However, in manual testing, feedback indicated that without additional UI or UX work, it could be confusing for users attempting to create sticky headers if they accidentally set a non-root block to sticky, or for example, a block _within_ a header template part to sticky. The decision was to scale back the feature to just the root blocks for 6.2 to allow more time to explore a good solution for nested blocks. See [the discussion in this Gutenberg issue](https://github.com/WordPress/gutenberg/issues/47892#issuecomment-1423668583) for more context.

#### Why is positioning not available in global styles?

The position block support has been designed to be set within individual block attributes, as positioning is typically set for an individual block, rather than globally for a particular block type. Therefore, within `theme.json` the position _setting_ can be used to opt in to using the position support, however a block's position styles cannot be set within global styles. It should instead be set at the individual Group block level, either within a site template, or a page or post.

#### Further reading

* Tracking issue: https://github.com/WordPress/gutenberg/issues/47043
* Initial PR that introduced the block support: https://github.com/WordPress/gutenberg/pull/46142
* Discussion on why sticky position is only available for root blocks for now: https://github.com/WordPress/gutenberg/issues/47892#issuecomment-1423668583
* MDN article for position: https://developer.mozilla.org/en-US/docs/Web/CSS/position
* Trac ticket for adding properties to `safe_style_css` filter: https://core.trac.wordpress.org/ticket/57504

---

@bph for visibility, here's a draft dev note for the sticky position block support.

bph commented:
Thank you so much @andrewserong I think this should be a stand-alone dev note, as it's a pretty neat feature. Would you have bandwidth to post it as a draft on the Make Blog to get it into the publishing queue for the field guide?

andrewserong commented:
Thanks @bph, a stand-alone dev note sounds good to me! Here's a draft on the Make blog for sticky position block support: https://make.wordpress.org/core/?p=103009&preview=1&_ppp=7494b4dec7

bph commented:
Thank you! @andrewserong
